tensorflow/core/kernels/diag_op.cc

```diff
@@ -129,24 +130,38 @@ namespace functor {
 template <typename T>
 struct DiagFunctor<CPUDevice, T> {
   EIGEN_ALWAYS_INLINE Status
-  operator() (const CPUDevice& device, const int64 size,
-                   const T* in, T* out) {
-    std::fill(out, out + size * size, T());
-    for (int64 index = 0; index < size; index++) {
-      out[(1 + size) * index] = in[index];
-    }
+  operator() (OpKernelContext* context, const int64 size,
+              const T* in, T* out) {
+    auto subDiag = [in, out, size](int64 start, int64 limit) {
```

Comment:
Perhaps add a 1-line comment that says this code is responsible for writing values in index range [start\*size, limit\*size).   (so it's clear that it is not in range [start, limit).